### PR TITLE
Fix OTELCPP_MAINTAINER_MODE

### PR DESCRIPTION
**Pull request by marcalff**
Fix OTELCPP_MAINTAINER_MODE

## Changes

Fix OTELCPP_MAINTAINER_MODE
- bump gcc to version 12
- bump clang to version 14
- bump googletest to 1.12.1 (1.10 causes new build warnings with upgraded gcc and clang)
- expand coverage on W3C tests
- fix new warnings reported

* [X] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -24,36 +24,36 @@ jobs:
         CC=/usr/bin/gcc-10 CXX=/usr/bin/g++-10 ./ci/do_ci.sh cmake.test
 
   cmake_gcc_maintainer_test:
-    name: CMake gcc (maintainer mode)
+    name: CMake gcc 12 (maintainer mode)
```
Comment (ThomsonTan) [inline .github/workflows/ci.yml:27]: Is gcc 12 and clang 14 installed by default in the build runner?

Comment (marcalff) [reply .github/workflows/ci.yml:27]: Yes, they are the latest compiler in the runner image:

https://github.com/actions/runner-images/blob/ubuntu22/20221127.1/images/linux/Ubuntu2204-Readme.md

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -24,36 +24,36 @@ jobs:
...
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v3
       with:
         submodules: 'recursive'
     - name: setup
       run: |
-        sudo CC=/usr/bin/gcc-10 CXX=/usr/bin/g++-10 ./ci/setup_cmake.sh
-        sudo CC=/usr/bin/gcc-10 CXX=/usr/bin/g++-10 ./ci/setup_ci_environment.sh
+        sudo CC=/usr/bin/gcc-12 CXX=/usr/bin/g++-12 GOOGLETEST_VERSION=1.12.1 ./ci/setup_cmake.sh
```
Comment (lalitb) [inline .github/workflows/ci.yml:35]: If gcc-12 is the default compiler with ubuntu-latest, do we need to set it explicitly here ?

Comment (marcalff) [reply .github/workflows/ci.yml:35]: The installed GCC compilers are: GNU C++ 9.5.0, 10.4.0, 11.3.0, 12.1.0.

Which one is the default, I don't know.

Naming explicitly the compiler we want is more robust and avoids surprises, should this list change again with a different ubuntu-latest image.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks

**Review by owent — APPROVED** [review]
LGTM

**Pull request merged**